### PR TITLE
Hotfix/tifr

### PR DESCRIPTION
**Pull request by mikeaclark**
These are fixes required to get the TIFR MILC clover rhmc benchmark working properly, as well as minor interface fixes.

* Renamed inconsistently named interface functions in `quda.h`
* Added new milc interface function `qudaResidentExtendedGaugeField` for creating an returning an extended gauge field from the persistent gauge field
* Disable reliable updates for pure double precision multi-shift solver in the milc interface (this matches the behaviour of the milc staggered interface)
* Add doxygen comments to some missing functions in `quda.h`
* Added support for backing up `gauge::FloatNOrder` objects for use in auto tuning
* Clover derivative computation now uses `gauge::FloatNOrder` structs and now supports reconstruction
* Added autotuning support to clover derivative and sigma trace kernels.

I think the risk in adding this functionality is minimal.  Although some of the quda interfaces have changed, these are only called by the milc interface (which is internal to QUDA) and by no other external application.  Thus I think this is fine to include in the 0.7.1 release.

One comment though: the additions made to `gauge_field_order.h` are incompatible with the changes that @nmrcardoso's gauge-fixing branch has made.  It's a trivial fix though, and I'll push this fix to the feature/gauge-fix branch once this pull request has been done.

### CODE REVIEW

**Comment by mathiaswagner** [comment]
I have not looked at the changes yet but about getting this in 0.7.1:
Apart from the slight risk (which is fine with me if my next if applies) I would only go for this if we do have a strong motivation. How urgent is this?

What tests has this passed? 

Anyway, I will have a look at the changes.

**Comment by mathiaswagner** [comment]
First thing on my mind when changing the milc interface: While it completely misses doxygen comments right now this is not what we want for the future. Please add some comments so that at least the new stuff has them,

Should things like the turned off reliable updates for pure double precision solves also be mentioned in the doxygen comments then? I think doxygen comments are use documentation and this is something that might be of interest.

File: include/quda.h [hunk 1]
```diff
@@ -635,31 +635,74 @@ extern "C" {
...
+   * @param mu mu direction
+   * @param nu nu direction
+   * @param dim array of local field dimensions
+   */
+  void computeCloverTraceQuda(void* out, void* dummy, int mu, int nu, int dim[4]);
 
-  void computeCloverTraceQuda(void* out, void* clover, int mu, int nu, int dim[4]);
-
+  /**
+   * Compute the derivative of the clover term
```
Comment (mathiaswagner) [inline include/quda.h]: missing `@param`, might also use '@sa' ?

**Comment by mikeaclark** [comment]
To answer the question on urgency, this is something that really should be in 0.7.1.  These changes are required for the TIFR clover rhmc code to work correctly, this is an important application for a new Cray XC30 installed in Mumbai.  I want them to be able to use a golden version of QUDA without having to rely on some development branch.

File: lib/milc_interface.cpp [hunk 2]
```diff
@@ -878,14 +878,25 @@ void qudaInvert(int external_precision,
 
 #ifdef GPU_CLOVER_DIRAC
 
-void* qudaCreateExtendedGaugeField(void* gauge, int geometry, int precision)
+static inline void* createExtendedGaugeField(void* gauge, int geometry, int precision, int resident)
 {
   qudamilc_called<true>(__func__);
```
Comment (mathiaswagner) [inline lib/milc_interface.cpp]: Just wondering whether we should us `qudamilc_called<>` only in exposed functions. 
However, we can keep it this way for now and I will create a separate cleanup issue and we can decide whether we also want to have the `qudamilc_called<false>` call before exiting for all exposed functions.

But as this is only verbose output it does not prevent the merge. It just made me aware of another inconsistency in the milc interface.

Comment (AlexVaq) [inline lib/milc_interface.cpp]: Mathias, are we having the developers call? Did I miss something? Did I understand wrong the time???

**Pull request merged**